### PR TITLE
Perf improvement: refine scheduling logic in shim to reduce num of events

### PR DESCRIPTION
**Pull request by yangwwei**
This is a replacement for https://github.com/cloudera/yunikorn-k8shim/pull/56. Refined the scheduling logic to reduce the num of events generated in each scheduler interval.

### CODE REVIEW

**Comment by yangwwei** [comment]
@TaoYang526 pls help to review this patch. See if this approach resolves the issue. Thanks!

**Comment by TaoYang526** [comment]
Thanks @yangwwei for updating this PR.
It's an awesome idea to let task handle state transition by itself, and it has clearly solved the problem. Good job!

File: pkg/cache/application.go [hunk 1]
```diff
@@ -203,6 +210,48 @@ func (app *Application) getTasks(state string) []*Task {
...
+		}
+	case states.Accepted:
+		ev := NewRunApplicationEvent(app.GetApplicationId())
+		if err := app.handle(ev); err != nil {
+			log.Logger.Warn("failed to handle RUN app event",
+				zap.Error(err))
+		}
+	case states.Running:
+		if len(app.GetNewTasks()) > 0 {
+			for _, task := range app.getTasks(events.States().Task.New) {
```
Comment (TaoYang526) [inline pkg/cache/application.go]: ApplicationInfo#getTasks will go through taskMap without readLock, may conflict when another thread is updating taskMap. I think we can query new tasks once like this:
```
newTasks := app.GetNewTasks()
if len(newTasks) > 0 {
   for _, task := range newTasks {
   ...
   }
}
```

Comment (yangwwei) [reply pkg/cache/application.go]: yeah, that's correct. let me fix that. thx

**Comment by yangwwei** [comment]
Thanks for the review @TaoYang526 , I have just merged this.

**Pull request merged**